hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py

```diff
@@ -1245,32 +1247,50 @@ def redistribute_block_pairs(self, chooser):
         to exceed some bound, lowering one of them requires raising the
         other. This pass enables that."""
 
-        block = chooser.choose(self.blocks, lambda b: not b.all_zero)
+        node = chooser.choose(
+            self.nodes, lambda node: node.ir_type == "integer" and not node.trivial
+        )
 
-        for j in range(block.index + 1, len(self.blocks)):
-            next_block = self.blocks[j]
-            if next_block.length == block.length:
+        # The preconditions for this pass are that the two integer draws are only
+        # separated by non-integer nodes, and have the same size value in bytes.
+        #
+        # This isn't particularly principled. For instance, this wouldn't reduce
+        # e.g. @given(integers(), integers(), integers()) where the sum property
+        # involves the first and last integers.
+        #
+        # A better approach may be choosing *two* such integer nodes arbitrarily
+        # from the list, instead of conditionally scanning forward.
+
+        for j in range(node.index + 1, len(self.nodes)):
+            next_node = self.nodes[j]
+            if next_node.ir_type == "integer" and bits_to_bytes(
+                node.value.bit_length()
```

Comment:
I didn't *want* to keep this condition, but otherwise the test for this fails: the ir tree for `@given(st.integers(), st.integers()` looks like `[integer {"max_value": 128}, integer, integer {"max_value": 128}, integer`], where the 128 draws are for choosing a forced endpoint. Without checking for equal sizes, we try to redistribute the endpoint draw to the actual draw, which of course does nothing.

As noted in the inline comment, there are plenty of other ways this pass can get tripped up! But we can look at improving or removing it later.

Reply:
Hmm.  I'm tempted to modify the forced-endpoint logic so that it doesn't do this, e.g. by using the `weights=` parameter instead of making two draws.  Thoughts?

Reply:
ah, perfect! Let's do that. Having an ir-strategy correspond to two ir nodes felt bad,

Reply:
ah, problem...we quickly run into our self-imposed width limit of 255 for weights. I imagine this was imposed because `Sampler` performance blows up otherwise. But we also definitely don't want to be constructing the 2**32 element list just to say "uniform everywhere except the endpoints", for memory reasons, among others (potentially float loss).

Maybe we need a better structure for expressing `weights`, which allows us to specify "upweight just a few elements from this very large range"? Potentially specifying start and end indices where the weighting should apply uniformly to that range, instead of one weight per element.

Reply:
I like that - new interface idea: `weights=` takes a mapping of {value: weight}, len <= 255 and 0 <= sum-of-weights <= 1.  Then we use a Sampler to pick either a value, or the remaining probability mass which means "pick according to unweighted".

Reply:
I think this is the right path forward. But it's fiddly to do so while maintaining correct invariants about forced and children count. Would you object to leaving it for a future PR? I suspect we may have some back and forth on this change.

Reply:
Yeah, just write a note in #3921 and we can get to it later 👍